### DISCUSSION TITLE
Segregated multi-site support

### DISCUSSION THREAD

**Discussion by DanielSwain** [wagtail/wagtail#7521] [Ideas]
The current multi-site support in Wagtail assumes that all sites are for the same organization.  If I wanted to create sites for multiple organizations of the same type (for instance, Chambers of Commerce) on a single Wagtail installation, then the images and documents uploaded by one organization are visible to all.  Also, the pages of all organizations are visible to everyone else in the Wagtail Explorer.  Creating multiple sites for organizations of the same type on a single Wagtail installation would provide the capability to view statistics across multiple organizations and to promote cross-organization communication among members (and probably other interesting things as well).  Please consider implementing segregated multi-site support in which users who are assigned editing permissions per site can only view the pages and assets of that site.

**Comment by DanielSwain** [comment]
Looks like Django sites framework's [CurrentSiteManager](https://docs.djangoproject.com/en/2.0/ref/contrib/sites/#the-currentsitemanager) might be useful.  Perhaps the requested functionality could simply be implemented as an option in Settings.

**Comment by gasman** [comment]
Hi @DanAtShenTech - thanks for the suggestion.

This is indeed something we're keen to implement, and there are various strands of work going towards this (#2869, #4206, #4207) - although it looks like we don't currently have an open issue for the feature as a whole since #2463 was closed, so I'll leave this one open.

It's possible to achieve most of this through user groups and permissions - in particular, a user who's been assigned permission over a single sub-site won't see other sites in the explorer. The main thing that's missing at the moment is introducing a 'choose' permission level so that we can restrict the pages shown in the 'choose a page' popup. Note that we want to avoid tying our model of limited-access users too closely to the concept of 'sites', since we want to keep things flexible enough to allow for other arrangements - for example, an organisation might want to enforce this kind of compartmentalisation between sections of a site, without having a distinct domain name for each one.

**Comment by bahoo** [comment]
> in particular, a user who's been assigned permission over a single sub-site won't see other sites in the explorer.

Could you elaborate on this? I'm having trouble finding any clear avenues for assigning permissions to a site. Thank you! :)

( +1 to fleshing this feature out further, also 👍 )

**Comment by mjlabe** [comment]
My friend and I hacked together something that sort of worked:

https://github.com/mjlabe/bwagtail-cms

We ended up filtering via SQL:

https://github.com/mjlabe/bwagtail-cms/blob/master/bwagtail/bwagtail/get_user_sites.py

And overriding the views for switching the site:

https://github.com/mjlabe/bwagtail-cms/blob/master/bwagtail/wag_custom/views.py

https://github.com/mjlabe/bwagtail-cms/blob/master/bwagtail/wag_custom/forms.py

**Comment by DanielSwain** [comment]
[This video](https://www.youtube.com/watch?v=RKSwjaZKXL0) might be useful in implementing this feature.

**Comment by bahoo** [comment]
I ran a bit farther with @mjlabe's implementation, and made a few small updates:

- updated `edit` and `edit_current_site` to match updated arguments ( circa 2.6.2 )
- monkeypatches / re-sets the view function instead of requiring re-registering all the Wagtail URLs ( though this approach is downright disgusting lol and I would love to hear a cleaner solution here )
- removes the references to `django-currentuser`

Not production ready, but if this is a helpful stepping stone in getting someone else toward a workable prototype, here you go:

https://gist.github.com/bahoo/5f073c15d0e3ff742ab14ecc5a671de9

**Comment by ababic** [comment]
While the tree-based permissions of the page tree and collections allow a great deal of flexibility in defining how multi-tenancy should work for a project - I can't help bit think we're missing something to pull it all together. 

I would like to propose a new `CMSInstance` model to fulfil this role - which would represent an 'instance of the CMS' accessed by editors (as opposed to a `Site` which is accessed by front-end users). And like a `Site`,  a `CMSInstance` would be identifiable by the domain of the current request, but it would only apply to the CMS (which might be at an entirely different domain to any of the `Site` objects).

Like we did for images and documents when Collections were introduced, we could create a single `CMSInstance` by default and associate all `Site` objects with it (via a new `ForeignKey` field on the `Site` model), and things like the site switcher in `wagtail.contrib.settings` (and perhaps the site list itself) would be limited to sites associated with the current `CMSInstance`.

All views and choosers would use the current `CMSInstance` to filter out anything from other instances, and automatically associate content with the correct instance. And, like we do with `wagtail.search.models.Indexed`, we could include a model mixin (e.g. `CMSInstanceSpecific`) that developers could use to create models with content unique to each instance.

We'd probably also need to provide a version of `CMSInstanceSpecific` to help create tenant-specific user models, (username would have to be unique per `CMSInstance`, rather than completely unique), which would be important for managing users / configuring admin permissions on a per-instance basis (admin authentication would need a rejig too). 

A few other things that would need thinking about are:
- What the relationship between `CMSInstance` / `Collection` / `Image` / `Document` should be
- What we'd do about tags (each `CMSInstance` should probably have it's own tag set/cloud)
- How would non-CMS users be associated with a `CMSInstance` - maybe indirectly via `Site`?
- Whether there would be a UI to manage this, or whether (like Collections), developers would manage `CMSInstance` objects outside of Wagtail.

Interested to hear what folks think about this general approach? @mjlabe @bahoo @DanielSwain Does this sound like something that would fit your use-cases? If feedback is good, I'll work up these ramblings into a proper RFC.

**Comment by mjlabe** [comment]
It sounds great to me. I think it would be best to have dropdowns to assign
a site to a CMS instance. Even if Django admin were left as is, I still
think it would work for me. The Django admin site is for true admins with
access to everything and the wagtail admin would work on a per CMSInstance
basis.

**Comment by DanielSwain** [comment]
To add to this discussion, I am including here references to my [two comments](https://github.com/wagtail/wagtail/pull/5375#discussion_r301572535) from July 9, 2019.  I also plan to add some other thoughts here as soon as I have time to write them up.

**Comment by carlb** [comment]
@ababic did your proposal ever make it into an RFC (I looked [here](https://github.com/wagtail/rfcs/pulls and couldn't find it)? My group (in a clinical informatics lab in Boston) are just (i.e. the past 2 weeks) getting started with Wagtail to deal with a sudden surge of content related to the pandemic across our network of hospitals. Loving it so far, but the enhancement discussed here sounds like it might be ideal for our use case now and longer term. In our case it's not different Chambers of Commerce (the example used by @DanielSwain above), but different hospitals, academic centers, and other organizational units that we support that may have similar content, needs, but need to be segregated in the admin UI.

Would love to contribute to the discussion around this and pick a direction for our urgent need that is compatible with community plans to address this.

**Comment by ababic** [comment]
Hi @carlb. Thanks for posting! It just so happens I implemented something like the above for a client project recently. It made a few things clearer in my mind, but many elements of the solution were probably a bit too bespoke to apply to many projects (otherwise I would have open-sourced it). 

I haven't put together an RFC yet, as the issue is so multi-facetted that it probably demands 3 or 4 RFCs to tackle the various different aspects, and I simply haven't had the time to put them together (although I'd certainly still like to). 

The `CMSInstance` idea (I went with `CMSSite` in the end) definitely has legs, but I only managed to make it work with each site being administered at a separate (single) CMS domain, e.g. 
- `cms.foo.com` manages content for `foo.com`
- `cms.bar.org` manages content for `bar.org` 

Making things completely segregated is really quite a challenge, I've found. If you want true segregation between each site, I would highly recommend investing in infrastructure instead, and simply host the same codebase multiple times (especially if free hosting plans will do, and the sites are well-suited to upstream caching with something like Cloudflare). Having separate instances makes environment configuration and resource usage monitoring easier, and not depending on a single environment will make things more resilient overall.

If you do need to have cross-talk between sites (or some kind of 'overview'), it should be possible to implement REST/Graph API endpoints to achieve that. And if you're concerned about users having to manage accounts/passwords for multiple sites, a basic SSO implementation can help with that.

**Comment by DanielSwain** [comment]
I am still very hopeful to see this feature come to pass.  In preparation for this eventuality, I want to mention my comment [here](https://github.com/wagtail/wagtail/pull/5375#discussion_r301638555) about use of a `themes` folder.

**Comment by ababic** [comment]
@DanielSwain Broadly speaking, I would say the remaining pieces of the puzzle are:

- [ ] Support for a `can_choose` permission for Pages, and respect throughout (e.g. in modeladmin)
- [x] Support for a `can_choose` permission for Collections
- [x] Support for nested collections in the Wagtail UI ([RFC](https://github.com/wagtail/rfcs/pull/48))  
- [ ] Support for site-specific collections (each site can specify it's own root collection, and that collection will be selected by default in various places when working 'on that site')
- [ ] Support for "Swappable site identification policies" - Allowing developers to customise how `Site.find_for_request()` works (e.g. by the CMS domain being used) 
- [ ] A `CMSSite` model, which would allow you to configure which sites are administered at which domains (and other details: such about who is allowed to sign into which `CMSSite`)
- [ ] Where multiple domains are used to edit different sites - recording in a user's profile which `CMSite` they last accessed or tend to access more regularly, so that email communications including links to the CMS can include domain that is relevant to the user and/or context of the email.
- [ ] Support for site-specific (delegated) user management (non-superuser user administrators can only manage user 'roles' related to their site, and cannot promote themselves or anyone else to a state that is more permissive than their own)
- [ ] Support for site-specific 'tag clouds' (sharing the same model, but serving different results depending on the 'site being administered')
- [ ] Adding some mixin models / utilities to help implement site-specific models and manage them via Wagtail
- [ ] A way to associate groups with sites, and limit their visibility depending on that association (in the main list, and also where roles are selected for a user)
- [ ] Some way to manage user permissions for multiple sites that doesn't require manual duplication/management of lots of site-specific groups
- [ ] Site-specific theming (perhaps a native version of https://github.com/moorinl/wagtail-themes)

Even with all these things in place, I think Wagtail might still have to recommend separate hosting as the favoured way to segregate sites (Where the number is manageable... e.g. approx 15 or fewer), not least because all it takes is for someone to make the Django admin area available (or some other Django app that doesn't care about Wagtail's site-specific view logic), and you immediately have important security and privacy implications.

**Comment by robslotboom** [comment]
To handle some restrictions I created the folowing models.
Rootpage is the toplevel – independent of any site – to add home pages to.

```python
class ProxySite(Site):
    
    objects = ProxySiteManager() < see below
    
    class Meta:
        proxy = True

    @property
    def home_page(self): 
        return self.root_page

class RootPage(Page): < toplevel
    parent_page_types = []
    subpage_types = ['cms.home_page')]
    is_movable = False
    is_copyable = False
    is_deletable = False
    is_unpublishable = False

    @property
    def can_add_homepage(self):
        return self.get_children().count() < MAX_SITES

class HomePage(Page):

    max_count = MAX_SITES
    is_movable = False
    is_unpublishable = False
    @property
    def is_copyable(self):
        return self.get_parent().specific.can_add_homepage

    @property
    def is_deletable(self):
        return self.get_parent().get_children().count() > 1

    subpage_types = [
            'cms.search_page',
            'cms.default_page',
            'cms.album_page',
            'cms.news_page'
        ]

class DefaultPage(Page): 
    
        parent_page_types = [
            'cms.home_page',
            'cms.default_page'
        ]
        subpage_types = [
            'cms.default_page',
        ]

class ProxySiteQuerySet(models.query.QuerySet):

    def default(self):
        try:
            return self.get(is_default_site=True)
        except:
            return
    
    def not_default(self):
        return self.filter(is_default_site=False)
    
    def default_or_first(self):
        return self.default() or self.first()
    
    def get_user_sites(self, user):
        user_perms = UserPagePermissionsProxy(user)
        if user_perms.can_edit_pages():
            site_pks = list(page.get_site().pk for page in user_perms.editable_pages().all())
            site_pks = list(sorted(set(site_pks)))
            return self.filter(pk__in=site_pks).distinct().all()
        
        return self.none()
        
    def get_current_user_site(self, user, site):
        user_site = self.none()
        user_sites = self.get_user_sites(user)
        try:
            user_site = user_sites.get(hostname=site.hostname)
        except ObjectDoesNotExist:
            user_site = user_sites.first()
        return user_site

    def get_sites_for_hostname(self, hostname, port):
        return self.annotate(
            match = Case(
                When(hostname = hostname, port = port, then = MATCH_HOSTNAME_PORT),
                When(hostname = hostname, is_default_site = True, then = MATCH_HOSTNAME_DEFAULT),
                When(is_default_site = True, then = MATCH_DEFAULT),
                default=MATCH_HOSTNAME,
                output_field=IntegerField()
            )
        ).filter(
            Q(hostname=hostname) | Q(is_default_site=True)
        ).order_by(
            'match'
        ).select_related(
            'root_page'
        )

class ProxySiteManager(models.Manager):

    """ Is in wagtail’s SiteManager """
    def get_by_natural_key(self, hostname, port):
        return self.get(hostname=hostname, port=port)

    def get_queryset(self):
        return ProxySiteQuerySet(self.model, using=self._db)

    def default(self):
        return self.get_queryset().default()
    
    def not_default(self):
        return self.get_queryset().not_default()

    def default_or_first(self):
        return self.get_queryset().default_or_first()

    def get_user_sites(self, user):
        return self.get_queryset().get_user_sites(user)
    
    def get_current_user_site(self, user, site):
        return self.get_queryset().get_current_user_site(user, site)
    
    def get_sites_for_hostname(self, hostname, port):
        return self.get_queryset().get_sites_for_hostname(hostname, port)
```

Editors are linked to one or more home page using group and permissions.

And some hooks prevent unwanted actions.

```python

@hooks.register('before_delete_page')
def before_delete_home_page(request, page):
    if request.method == 'POST':
        return

    home_page = request.proxy_site.home_page.specific
    specific_page = page.specific
    if specific_page.__class__ == home_page.__class__:
        if request.user.is_superuser:
            messages.warning(
                request,
                _('This is the {home_page} of {site_name}. When deleted the site will be removed to.').format(
                    home_page  = home_page._meta.verbose_name.lower(),
                    site_name = request.proxy_site.site_name
                )
            )
        else:
            messages.error(
                request,
                _('You are not allowed to delete the {home_page}.').format(
                    home_page = home_page._meta.verbose_name.lower()
                )
            )
            return redirect('wagtailadmin_explore', page.id)
            

@hooks.register('before_copy_page')
def before_copy_home_page(request, page):

    home_page = request.proxy_site.home_page.specific
    specific_page = page.specific
    if specific_page.__class__ == home_page.__class__:
        if request.user.is_superuser:
            root_page = home_page.get_parent().specific
            if root_page.can_add_homepage:
                messages.warning(
                    request,
                    _('Please note that you are about to copy the {home_page}.').format(
                        home_page = home_page._meta.verbose_name.lower()
                    )
                )
            else:
                messages.error(
                    request,
                    _('The root page doesn’t allow you to copy the {home_page}.').format(
                        home_page = home_page._meta.verbose_name.lower()
                    )
                )
                return redirect('wagtailadmin_explore', page.id)
        
        else:
            messages.error(
                request,
                _('You are not allowed to copy the {home_page}.').format(
                    home_page = home_page._meta.verbose_name.lower()
                )
            )
            return redirect('wagtailadmin_explore', page.id)
```

**Comment by robslotboom** [comment]
Unfortunately the above code shows an errormessage in the navigator due to 404 for the rootpage raised by the api:

```
class PagesAdminAPIViewSet(PagesAPIViewSet):

    def get_queryset(self):
        queryset = super().get_queryset()
        # Hide root page
        # TODO: Add "include_root" flag
        queryset = queryset.exclude(depth=1).specific()
        return queryset
```

**Comment by bahoo** [comment]
FWIW / if it's helpful — folx on Postgres, I've had some success with Django Tenants, until ( / if ) Wagtail does its own in-house so to speak. https://github.com/django-tenants/django-tenants

**Comment by awhileback** [comment]
FWIW, I would echo Matt's comments above by example with my current project's use case:

Having discrete "sites" but not discrete users is incidentally very good for my present use case.  I'm building a project for media distribution, and one of the benefits of Wagtail's extra layer on top of Django is the ability to easily implement syndication across multiple sites (in that a user signed up for one site is signed up for all sites, and third party payment models like Stripe built on top of user permissions for multiple tiers of content across multiple sites "just work").

Not every thing under the sun is a site hosting platform intended to provide discrete websites for discrete users.  And for those where that is the use case, what can be coded in Python that's easier than deploying another container in the same pool of servers and another postgres schema in the same database deployment/instance?  I assure you that dockerfiles and ansible scripts are easier to write than extending the functionality of a framework ;).

**Comment by vsalvino** [comment]
I agree with @awhileback ... The general idea with multi-site is that the various sites are going to share users, data, templates, code, etc. So trying to segregate them seems to go against that idea.

It sounds like the OP is trying to build a product, specifically for hosting his customers (chambers of commerce). We also went down this road early on, and realized that mutli-site anything (Wagtail, WordPress, etc.) is the wrong approach (at least for our needs). Something of that magnitude is going to require a lot of custom development.

The more ideal use-case that comes to mind is use of Wagtail in a corporate environment where there may be separate brands used in a multi-site (e.g. MegaCorp has subsidiaries SmallerCorp and OtherCorp - both of which share similar website designs). In that case an employee of SmallerCorp should have some segregation from ability to edit OtherCorp, which the Wagtail permissions seem to already do a good job of. In this scenario SmallerCorp should not be able to delete/edit media, pages, etc. from OtherCorp, but I still have not seen any real-world example where they need total isolation between them. Any organization that needs total isolation between sites almost always ends up with totally different features, designs, etc. between the sites anyhow, and is happy to have two separate codebases (because at that point the developers may also need to be isolated!).

**Comment by cnk** [comment]
> Any organization that needs total isolation between sites almost always ends up with totally different features, designs, etc. between the sites anyhow, and is happy to have two separate codebases (because at that point the developers may also need to be isolated!).

We have definitely had people try to push us to add features or make design changes to accommodate their preferences. And in some instances we have caved to pressure and added features that we place behind feature flags so we can offer them on only one or two sites. But in general Caltech has been able to hold the line on only adding features that can be offered on all sites. This is mainly because "it is free" is a powerful incentive for academics. And we are able to offer sites for free because the marginal cost of a new site in a multitenant system is close to zero. 

I would like to find time to [revisit Andy's list](https://github.com/wagtail/wagtail/discussions/7521#discussioncomment-1334317) in this discussion and make some RFC proposals. But in the mean time, I have written up the [patches we have implemented to enforce multitenancy](https://cynthiakiser.com/blog/2023/11/01/multitenancy-with-wagtail.html).

**Comment by ababic** [comment]
I spent some time 'a year or so' ago doing a more thorough write-up about [how true/full Multitenancy might work in Wagtail](https://docs.google.com/document/d/1z8yvYcrckf8yIt0wknKBI1aXrGd2VuzXn-lnDFjB8Hc/edit?usp=sharing), and some more on a POC to get a feel for the code changes that would be required to support my ideas.

Upon presenting the changes to the core team, we concluded that, ultimately, the number of practical use-cases  for native multitenancy are so few that the added project complexity for such an approach would be too burdensome, and not something they would be willing to persue.

**Reply by ababic** [reply]
This said, there could still be legs in exploring a "Segregated multi-site" approach (directly linked to sites), instead of a full-on "Tenanted" one (implemented as a separate layer above multi-site).

**Comment by thibaudcolas** [comment]
Revisiting this in late 2024, I’m unclear how much the needs expressed here could be addressed with "better multi-site support" (which it feels like we could get to with gradual improvements to the current implementation), and how much there are other needs that would require further segregation? Though there’s a risk of a "no true scotsman" here, where true multi-tenancy is always one step further than we are at.

For example in the needs initially expressed by @DanielSwain in 2018:

> the images and documents uploaded by one organization are visible to all. Also, the pages of all organizations are visible to everyone else in the Wagtail Explorer.

I don’t think that’s true (anymore?), with permissions that can be set both for visibility of the page tree, and of image / document collections.

@ababic’s [list from 2020](https://github.com/wagtail/wagtail/discussions/7521#discussioncomment-1334317) contains a lot of things that are definitely not supported right now (for example site-specific users and groups).

---

Personally what I think would be really helpful at this stage of the conversation is:

1. **A multi-site demo site**, so more of us can more easily get up to speed with current features, and therefore gaps. I started with my [bakerydemo-nextjs](https://github.com/thibaudcolas/bakerydemo-nextjs) project, as we also need a "headless" demo site, and multi-site tends to be more relevant for headless builds.
2. **A matrix of all Wagtail features with their respective degree of current multi-site support**, and an indication of the need for more fundamental changes for multi-tenant support. Could be based on Andy’s list, or perhaps something like my [Wagtail 5.1 UI overview](https://docs.google.com/spreadsheets/d/1FMSA_BI3ZvkeAvuaIL2QtqRTgMNwz_vhfKBeyx2Onnk/edit?gid=1962441802#gid=1962441802).

We have [mutli-tenancy improvements](https://github.com/wagtail/roadmap/issues/74) on our roadmap in the "future" area, so it’d be great to express clearly what gaps / improvements there are in this space.

**Comment by NotANormalNerd** [comment]
@thibaudcolas 
I was just about to do a small writeup of our working multi-tenancy setup. We are using it in production and have found some weak point in wagtail that we had to patch, but all in it works really well. There are some things that still sometime bug out but all in all we're really happy and so are our users.

### Wagtail multi-tenancy with [django-tenants](https://github.com/django-tenants/django-tenants)

I already used [django-tenants](https://github.com/django-tenants/django-tenants) in some other projects, so that was what I was familiar with, so that was what we are using. django-tenants is postgres specific and we switched from mysql to postgres, not only for that but also because Django and postgres are working better together.

django-tenants allows to have a public tenant and normal tenants. For our headless stack we put the API on the public tenant and did some tenant matching ourself and the admin UI completely on the normal tenants (currently only two, but more to come). Everything is separated on a database level by postgres search paths. But you could decide to have shared models/apps. We also used the `SubfolderMiddleware` instead of the domain middleware which lead to a few other problems. If you use the normal `TenantMainMiddleware` there should be less problems.

We needed to do some changes to the wagtail codebase so we wouldn't have shared state:

* Remove the `@cached_property` in `wagtail/admin/menu.py` and `wagtail/admin/search.py` because it was leaking urls into the tenants. 
* Now that I look at the code, actually nothing more. 

I wouldn't want to complicate the already complex codebase any further with custom code implementations. As far as I can see there is already multi-site support (one admin ui, multiple sites, same db) and I would rather focus on supporting a framework like django-tenants. There is the drawback of only supporting postgres, but I feel this is a fair tradeoff and time could be invested more in documenting how to use django-tenants with wagtail and what to watch out for.

**Reply by cnk** [reply]
@NotANormalNerd Very interesting. Would you have time to write up your system as a blog post? I would be really curious to see what you needed to do to make this work - especially the public vs normal tenant items. 

I think this could be a promising avenue for groups with 2-30ish tenants. But you have to run `migrate` once per tenant, right? So it would be unwieldy for my install with >500 sites.

**Reply by NotANormalNerd** [reply]
@cnk  Yes, I can do that. It's not rocket science, there are a few things here and there you have to be cautious, but all in all it provides an amazing level of customizability. 

Yes, in regards to migrations, you are absolutely right. Migrations are a challenge, but django-tenants has some optimisations as well as well as some operational options that you could choose. For example: [migrate_schemas in Parallel](https://django-tenants.readthedocs.io/en/latest/use.html#migrate-schemas-in-parallel)

**Reply by thibaudcolas** [reply]
Thank you @NotANormalNerd! I think we’d be up for changes in Wagtail itself to make it more compatible with django-tenants if it’s not too much work 🤔 My hesitation is that it being Postgres-only feels like a bit underwhelming.

Do you have any data shared between tenants? Personally I’d enjoy a bigger write-up where you explain why you arrived at this setup 😛

**Reply by NotANormalNerd** [reply]
@thibaudcolas Thats nice to hear 😄 

Yes, that is a drawback indeed, but django-tenants builds on functionality specifically provided by postgres, but that makes it quite elegant and easy to use.

Currently we don't have any data shared by the tenants (aside from the django-tenant data), but data can be shared per application, so you could share users or some kind of events database, whatever your heart desires. You will have to build the project with that in mind because obviously tenant data can point to public data, but not the other way round, otherwise you have integrity errors. This would also mean that you couldn't share a certain pagetype only, because they point to the Page Model.

I will write a show&tell here in Github next week, with some more info and reasoning. Some of it is really just: I have used it in another SaaS project successfully with 20 tenants, so I was familiar with it, but i will write that down.

**Reply by NotANormalNerd** [reply]
@thibaudcolas @cnk I have written down my approach here: https://github.com/wagtail/wagtail/discussions/12653

I don't know if this is what you wanted to know, if not, please feel free to ask. I would also be happy to hop on a call and show some of our code, if you are interested. I just can't share it here.

**Reply by thibaudcolas** [reply]
Great stuff!

**Comment by suxilog** [comment]
Hi @thibaudcolas I have project (Wagtail framework developed a year ago), designed to support multi-tenant websites. While it was originally intended for commercial use, due to time and resource constraints, I wasn't able to continue its maintenance. Now, I am open-sourcing it, hoping it can inspire others. Will upload later.

### Project Overview:
The system supports multiple tenants (in the form of organizations), with each organization being able to independently manage its site content. By combining the powerful features of **Wagtail** and **Django**, it provides flexible and secure multi-tenant management.

### Key Features:

1. **ExtendedSiteMiddleware**  
   This middleware checks if the request matches an extended site, ensuring that only active sites can be accessed and providing a custom 404 page.

2. **SiteSettingsAdmin**  
   Uses Wagtail's ModelAdmin to manage site settings, filtering the data displayed based on user permissions, ensuring that admins can only access settings for their respective sites.

3. **Granular Permission Control (CommonPermissionHelper and ExtendedSitePermissionHelper)**  
   Defines permission management mechanisms for different models, ensuring that only authorized users can access and modify the respective site settings, meeting the security needs of a multi-tenant environment.

4. **Custom Wagtail Page and Media Choosers**  
   Customizes the page and media choosers to ensure that users can only access and manage content belonging to their own organization, enhancing data security.

5. **Organization Management (OrganizationAdmin)**  
   Centralizes the management of organization information through Wagtail's admin interface, streamlining the organization management process.

6. **User Management (manage_users view)**  
   Allows organization admins to create new users and assign roles, ensuring that user management actions are restricted to authorized administrators only.

7. **Custom Site Settings Form**  
   Supports the dynamic generation of site-specific settings forms and tracks setting changes for traceability.

8. **robots.txt View**  
   Dynamically returns the `robots.txt` file content based on the current site's settings, controlling search engine crawler access.

9. **Page Filter & Search**  
   Supports multi-condition filtering for page searches, ensuring users can only see the content they are authorized to access.

10. **Celery Task to Execute Docker Commands (run_command_in_container)**  
    Automates command execution in Docker containers via Celery tasks, suitable for DevOps, CI/CD processes, and automation tasks.

11. **Dynamic CSRF Trusted Origin Setting (DynamicCSRFMiddleware)**  
    Dynamically calculates and sets the CSRF trusted origins for each request, ensuring security for cross-origin requests, especially important in a multi-tenant environment.

---

### Notes:
- Some of the code and comments in the project are in Chinese, but this does not affect the functionality, and you can use it as a framework for development.
- This open-source project offers full multi-tenant support and flexible content management features, making it suitable for reference or use in similar projects.

I hope this framework can inspire your multi-tenant website development!

**Reply by suxilog** [reply]
Project link: https://github.com/suxilog/Tentron

**Reply by thibaudcolas** [reply]
Very interesting, thank you for sharing! The [group permissions management](https://github.com/suxilog/Tentron/blob/aec5f3251308bc6350caaaf7ee1de33cfff915fa/organization/models.py#L708) is definitely something I’ll come back to refer to.

**Comment by thibaudcolas** [comment]
Discussions from core team meeting:

- Need to define the place of django-tenants in this ecosystem
- Work on snippet viewsets and permissions have been very helpful (hat tip @laymonage)
- Collection permissioning is a thing
- Two pain points from @cnk feedback:
  1. Figure out for a given page what site it belongs to? (Have to look at the whole tree)
    - A page could be part of a specific site, or a parent site. Relevant for permissions
    - @gasman: so far, have always assumed a page belonged to the "inner" / "deeper" site? 
  2. Attaching collections to sites ("root collection for specific site", or "collection for sites 1, 2, 3")
    - Again would help with permissions. 
- How about snippets? Make them more collection-y?
  - @cnk’s approach: site implementer adds a site field and move on
  - To make it easier: default mixin?
- Could we get more separation?
  - Related: SSO (LDAP) usage means more specific requirements on group management
- This vs. multi-instance: shared content

Possible next steps:

- Review what we could want to attach to sites: groups, collections, (snippets mixin)
- Consider how to make management of hostnames less tied to sites

**Comment by thibaudcolas** [comment]
Transferring message from @GabrielBogo:

Hi team. I've recently hit a scale-related performance issue, so I'm adding the details of what I found and how I worked around it for future reference.

---
### PROBLEM
I have a multi-site deployment with restricted permissions. Currently we have:

- 47 users
- 4 groups
- 171 sites
- 26 locales

... and granular permissions on a site-by-site basis across the 4 groups. Site-to-group permissioning is currently done via the Groups edit view, by giving the Groups full permissions to the root pages of each translation of each site.

This generates hundreds of `GroupPagePermissions`.

The problem is that upon opening the root-level Page List View, as well as using the search bar on it, requests are taking around 26 seconds to run when the user is not a superuser. For superusers, it's almost instantaneous.

I did some investigation, and it looks like permissioning is the problem. Particularly, the snippet below when run on the shell for non-superusers, takes the exact 26 seconds above, indicating that it has nothing to do with indexing (i.e., moving from Postgres to Elasticsearch as search backend would probably not solve the problem).

```
import time
from wagtail.permissions import page_permission_policy
from users.models import User

def time_explorable_instances(user):
    start = time.time()
    page_permission_policy.explorable_instances(user)
    end = time.time()
    print(end - start)

>>> user = User.objects.filter(is_superuser=False).first()
>>> time_explorable_instances(user)
>>> 26.550756454467773

>>> user_super = User.objects.filter(is_superuser=True).first()
>>> time_explorable_instances(user_super)
>>> 7.891654968261719e-05
```

If the number of sites increased to the point that the processing time reached 30s, my Heroku deployment would no longer have been able to serve the requests and the server would essentially crash. So it was a bit of a time bomb that I was handling.

### WORKAROUND

- Creating a parallel, simplified access control system. Basically, for each Site, I added a many-to-many field on its BaseSiteSetting containing the Groups that have access to that Site.

- Removing all GroupPagePermissions that were being used to control access to Page Trees. It looked like they were the culprit for the `explorable_pages` slowness. Now, all Groups only have access to Root.

- Using Wagtail Hooks to throw `PermissionError` before all major actions (before_edit_page, before_publish_page, etc).


Not great and it is ultimately a shadow-permission-control system, but at least I stopped the crashing for the time being without exposing anything to unauthorized users.

--

Looking forward to multi-tenancy support in the near future!

Thanks,
Gabriel
